docs/pipeline_semantics.md

```diff
@@ -0,0 +1,170 @@
+
+The usage of unbuffered channels between nodes and the absence of explicit buffers results in backpressure. This means
+that the speed of the destination connector dictates the speed of the whole pipeline.
+
+There is an implicit buffer that needs to be filled up before backpressure takes effect. The buffer is equal to the
+number of nodes in a pipeline - similar to how a longer garden hose holds more water, a longer Conduit pipeline can hold
+more messages. There are two exceptions to this rule:
+
+- Conduit is using gRPC streams to communicate with standalone connectors, which internally buffers requests before
+  sending them over the wire, thus creating another implicit buffer (we are aware of this
+  issue: [#211](https://github.com/ConduitIO/conduit/issues/211)).
+- Destination connectors are allowed to collect multiple records and write them in batches to the destination, which
+  creates a buffer that depends on the connector implementation.
+
+If there are multiple destinations the fanout node won't fetch a new message from the upstream node until the current
+message was successfully sent to all downstream nodes (this doesn't mean the message was necessarily processed, just
+that all downstream nodes received the message). As a consequence, the speed of the pipeline will be throttled to
+accommodate the abilities of the slowest destination connector.
+
+### Dead letter queue
+
+Messages that get negatively acknowledged can be rerouted to another destination called a dead letter queue (DLQ) where
+they are stored and can be reprocessed at a later point in time after manual intervention. If rerouting is set up and
+the message successfully reaches the DLQ the message will be acked instead. The user has the option to configure a DLQ
+that simply logs a warning and drops messages to achieve "at most once" delivery guarantees.
+
+### Pipeline stop
+
+A pipeline can be stopped in two ways - either it's stopped gracefully or forcefully.
+
+- A graceful stop is initiated either by Conduit shutting down or by the user requesting the pipeline to stop. Only the
+  source connector nodes will receive the signal to stop running. The source nodes will stop running and close their
+  outgoing channels, notifying the downstream nodes that there will be no more messages. This behavior propagates down
+  the pipeline until the last node stops running, draining any open messages in the process.
+- A forceful stop is initiated when a node stops running because it experienced an unrecoverable error. In that case,
+  the context that is shared by all nodes will get canceled, signaling to all nodes simultaneously that they should stop
+  running as soon as possible. Messages that are in the pipeline won't be drained, instead, they are dropped and will be
+  requested from the source again once the pipeline is restarted. The error returned from the first node that failed
+  will be stored in the pipeline and can be retrieved through the API.
```

Comment:
should we specify which API?

Reply:
I didn't go into detail here on purpose, it would be out of scope for this document. To be honest, I was reluctant even mentioning the API here, but didn't find a better way of saying that the error is stored and retrievable.

Reply:
agree